src/node/net.js

```diff
@@ -0,0 +1,1109 @@
+};
+
+Socket.prototype.resume = function() {
+  if (this.destroyed) return;
+  maybeStartReading(this);
+  return Duplex.prototype.resume.call(this);
+};
+
+Socket.prototype.read = function(n) {
+  if (this.destroyed) return;
+  maybeStartReading(this);
+  return Duplex.prototype.read.call(this, n);
+};
+
+Socket.prototype._read = function(n) {
+  if (this.connecting || !this._handle) {
+    this.once('connect', () => this._read(n));
+  } else if (!this._handle.reading) {
+    maybeStartReading(this);
+  }
+};
+
+// ======================================================================================
+// Destroy and reset
+
+Socket.prototype._reset = function() {
+  return this.destroy();
+};
+
+Socket.prototype.resetAndDestroy = function() {
+  // In Node.js, the resetAndDestroy method is used to "[close] the TCP connection by
+  // sending an RST packet and destroy the stream. If this TCP socket is in connecting
+  // status, it will send an RST packet and destroy this TCP socket once it is connected.
+  // Otherwise, it will call socket.destroy with an ERR_SOCKET_CLOSED Error. If this is
+  // not a TCP socket (for example, a pipe), calling this method will immediately throw
+  // an ERR_INVALID_HANDLE_TYPE Error." In our implementatin we really don't have a way
+  // of ensuring whether or not an RST packet is actually sent so this is largely an
+  // alias for the existing destroy. If the socket is still connecting, it will be
+  // destroyed immediately after the connection is established.
+  if (this.destroyed) return this;
+  if (this._handle) {
+    if (this.connecting) {
+      this.once('connect', () => this._reset());
+    } else {
+      this._reset();
+    }
+  } else {
+    this.destroy(new ERR_SOCKET_CLOSED());
+  }
+  return this;
+};
+
+Socket.prototype.destroySoon = function() {
+  if (this.destroyed) return;
+  if (this.writable) {
+    this.end();
+  }
+
+  if (this.writableFinished) {
+    this.destroy();
+  } else {
+    this.once('finish', this.destroy);
+  }
+};
+
+Socket.prototype._destroy = function(exception, cb) {
+  if (this[kTimeout]) {
+    clearTimeout(this[kTimeout]);
+    this[kTimeout] = undefined;
+  }
+
+  if (this._handle) {
+    this._handle.socket.close().then(
+        () => cleanupAfterDestroy(this, cb, exception),
+        (err) => cleanupAfterDestroy(this, cb, err || exception));
+  } else {
+    cleanupAfterDestroy(this, cb, exception);
+  }
+};
+
+// ======================================================================================
+// Connection
+
+Socket.prototype.connect = function(...args) {
+  if (this.connecting) {
+    throw new ERR_SOCKET_CONNECTING();
+  }
+  // TODO(later): In Node.js a Socket instance can be reset so that it can be reused.
+  // We haven't yet implemented that here. We can consider doing so but it's not an
+  // immediate priority. Implementing it correctly requires making sure the internal
+  // state of the socket is correctly reset.
```

Comment:
Should we add a nice error message for this case at least? maybe we already have it somewhere else in the code?

Reply:
That's already covered by the `this.destroyed` check below